### PR TITLE
make nx an optional dependency

### LINKED ISSUE #177: Make Nx an optional dependency
The `keyword!` function is now part of Elixir as `Keyword.validate!`.

For the tensor operations (to_tensor and from_tensor), we can add `@compile {:no_warn_undefined, Nx}`.

### PR DESCRIPTION
**Pull request by cigrainger**
This closes #177.

### CODE REVIEW

File: mix.exs [hunk 1]
```diff
@@ -32,7 +32,7 @@ defmodule Explorer.MixProject do
   defp deps do
     [
       {:ex_doc, "~> 0.24", only: :docs, runtime: false},
-      {:nx, "~> 0.1.0"},
+      {:nx, "~> 0.1.0", only: :test},
```
Comment (josevalim) [inline mix.exs:35]:
```suggestion
      {:nx, "~> 0.1.0", optional: true},
```

Comment (polvalente) [reply mix.exs:35]: Should be 0.2.0 actually, right?

Comment (cigrainger) [reply mix.exs:35]: TIL!

Comment (josevalim) [reply mix.exs:35]:
```suggestion
      {:nx, "~> 0.1", optional: true},
```

It is unlikely the API used by Explorer will ever actually change/break.

**Comment by josevalim** [comment]
@cigrainger you also need to change the mix.exs to say it requires Elixir ~> 1.13. CI already seems to use 1.13.

**Review by josevalim — APPROVED** [review]

**Pull request merged**